juju/main.py

```diff
@@ -0,0 +1,164 @@
+# This happens because the directory containing a script is added to
+# sys.path by the interpreter which gets resolved if it is a symlink.
+del sys.path[0]
+sys.path.insert(0, os.path.join(os.environ.get("JUJU_CHARM_DIR", ''), 'lib'))
+
+import juju.framework # noqa
+
+from juju.framework import ( # noqa
+    Framework,
+    Event,
+    BoundEvent
+)
+from charm import Charm # noqa
+
+# Reuse the same db file as used by charm-helpers.
+DB_PATH = os.path.join(
+    os.environ.get('CHARM_DIR', ''), '.unit-state.db')
+
+try:
+    from charmhelpers.core.hookenv import log
+except (ImportError, ModuleNotFoundError):
+    log = print
+
+
+def load_metadata():
+    with open('metadata.yaml') as f:
+        metadata = yaml.load(f)
+    return metadata
+
+
+def get_charm_endpoints():
+    """Get a list of endpoints from the charm metadata.
+    """
+    metadata = load_metadata()
+    charm_endpoints = list()
+    for endpoint_type in ['provides', 'requires', 'peers']:
+        endpoints = metadata.get(endpoint_type)
+        if endpoints:
+            charm_endpoints.extend(endpoints)
+    return charm_endpoints
+
+
+def setup_hooks():
+    """Set up hooks for supported events.
+
+    Whether a charm can handle an event or not can be determined by
+    introspecting which events are bound to it.
+    """
+
+    relation_events = list()
+    for endpoint in get_charm_endpoints():
+        relation_events.extend([
+            f'{endpoint}_relation_{t}'
+            for t in ['joined', 'changed', 'departed', 'broken']
+        ])
+
+    log(f'Discovered relation events from charm metadata: {relation_events}')
+
+    # Get core event names that CharmBase knows about.
+    core_events = [event_name for event_name, instance
+                   in juju.charm.CharmEventsBase.__dict__.items()
+                   if type(instance) == Event]
+
+    log(f'Discovered core charm events: {core_events}')
+
+    # Discover Juju events supported by the charm using reflection.
+    supported_juju_events = [
+        event_name for event_name in dir(Charm.on)
+        if (type(getattr(Charm.on, event_name)) == BoundEvent
+            and (event_name in relation_events or event_name in core_events))
+    ]
+    log('Discovered Juju events supported by the charm by'
+        f' introspecting it: {supported_juju_events}')
+
+    # Create symlinks to "install" which may be either a copy of main.py
+    # (which calls the entrypoint function) or a symlink to it. Note
+    # that it is important to avoid creating a recursive symlink.
+    supported_juju_events.remove('install')
+
+    for event_name in supported_juju_events:
+        event_hook_path = f'hooks/{event_name.replace("_", "-")}'
+        if os.path.exists(event_hook_path):
+            log(f'Removing an old symlink named {event_hook_path}')
+            os.unlink(event_hook_path)
+        log(f'Creating a new symlink named {event_hook_path} to hooks/install')
+        os.symlink('install', event_hook_path)
```

Comment:
Ideally the logic above would not remove/recreate the symlinks every time. It should instead know what's wrong (missing, or excessive) and fix it.